diff -up libhugetlbfs-2.16/Makefile.orig libhugetlbfs-2.16/Makefile
--- libhugetlbfs-2.16/Makefile.orig	2014-03-03 12:50:43.408107252 +0100
+++ libhugetlbfs-2.16/Makefile	2014-03-03 12:52:01.070230134 +0100
@@ -29,7 +29,7 @@ INSTALL = install
 
 LDFLAGS += -Wl,-z,noexecstack -ldl
 CFLAGS ?= -O2 -g
-CFLAGS += -Wall -fPIC
+CFLAGS += -Wall -fPIC -fstack-protector-strong
 CPPFLAGS += -D__LIBHUGETLBFS__ -DFORTIFY_SOURCE
 
 ARCH = $(shell uname -m | sed -e s/i.86/i386/)
